### ISSUE TITLE
Authenticate method should cache token response

### ISSUE THREAD

**Issue by tpazderka** [mozilla/mozilla-django-oidc#283]
`OIDCAuthenticationBackend.authenticate()` should cache the result of `OIDCAuthenticationBackend.get_token()` and `OIDCAuthenticationBackend.get_userinfo()`.

If multiple backends based on `OIDCAuthenticationBackend` are present in settings and the first one present is not able to authenticate the user (for example `verify_claims` returns False for missing claims), the second backend should reuse the already obtained data rather than trying to get them again.

This leads to code reuse and an error.

**Comment by johngian** [comment]
I don't thing that this is part of the scope of this library. Have you seen any performance issues that might be solved with caching in that level? It sounds more of something that should be implemented per case (eg. override ``verify_claims`` and cache results based on `sub`).

Closing this as wontfix.

**Comment by tpazderka** [comment]
The caching is not for performance reasons, but because the second backend gets an error from the OP since the code was already used by the first backend.

**Comment by johngian** [comment]
Thanks for the clarification @tpazderka. That indeed sound as a plausible scenario. Do you have a real world example of using multiple OIDC backends and expecting to work with a single access token?

**Comment by johngian** [comment]
We have a project with a similar setup that uses different providers so it works out fine.

**Comment by tpazderka** [comment]
Yes, our internal systems rely on single OP but different parts do need different sets of attributes. On the other hand, we want users to be authenticated if they come from a different part of our systems so we need to have the backends in settings.

**Comment by johngian** [comment]
I think what we can do to keep the scope of the project small but also workaround that scenario is to move all the OP calls to different methods, then library users can customize with caching logic for example.

```
def <op_api_call>(*args, **kwargs):
    try:
        call OP
    except:
        use cached values
    return value
```

**Comment by johngian** [comment]
Thoughts ?

**Comment by tpazderka** [comment]
Yes, it is possible to do it by overriding `get_token` on a user level, but my idea was to have something right of the bat.
